### ISSUE TITLE
ResolverCachePlugin: Allow dynamic multiple time resolve of the same path

### ISSUE THREAD

**Issue by Abdillah** [webpack/webpack#16909]
## Feature request
In Webpack 4, resolver resolve path multiple time (no cache), therefor plugin like this works: https://www.npmjs.com/package/dynamic-resolve-webpack-plugin. But in Webpack 5, ResolverCachePlugin automatically snapshot a resolve, this can't be modified without affecting other resolve.

Therefor, I am requesting ResolverCachePlugin to receive a function to determine whether the file need to be resolved again or not.

**What is the expected behavior?**

```
{
        resolve: {
            plugins: [
                new DynamicAliasResolvePlugin({
                    alias: "@template",
                    dynamic: (request, alias) => {
                        return request.context.issuer.replace(/\.js$/, '/template/');
                    },
                }),
            ]
        },
}
```

Using it like these on multiple files:

page-a.js
```
// Resolve to "./page-a/template/"
import Template from "@template/homepage";
```

page-b.js
```
// Resolve to "./page-b/template/"
import Template from "@template/homepage";
```

Two of the paths will be resolved on different path and not cached. 

**What is motivation or use case for adding/changing the behavior?**
The main motivation is the use of plugin such as DynamicAliasResolvePlugin. Currently, only `page-a.js` get resolved and `page-b.js` will use the previous resolve which led it to be unusable.

**How should this be implemented in your opinion?**
There are two way to implements this,

1. Add more options to `snapshot.resolve` that will be used to determine whether a file path need to be cached or not.
2. Pass and get config to resolver plugin whether the resulting resolve should be snapshot or not.

**Are you willing to work on this yourself?**
yes

**Comment by alexander-akait** [comment]
Sorry, I don't think it is possible, the resolver doesn't know nothing about cache and for me it is an unliked solution, sorry, I strongly recommend to migrate on something like `#one/@template/homepage` and ``#two/@template/homepage`, even more -  it will be faster and reduce your build time (you will not spend time on extra resolving)

**Comment by Abdillah** [comment]
I see.. 

May I ask if somehow I can reconstruct the problem space.

In which part of Webpack does a path resolved (e.g. `@page/function` resolved into `./page-a/function`) and kept (another `@page/function` later will resolved to  `./page-a/function` as well)? I think my deduction above that "ResolverCachePlugin does these caches" is wrong.

**Comment by alexander-akait** [comment]
> May I ask if somehow I can reconstruct the problem space.

Is it possible to rewrite imports? 

> In Webpack, which part makes a path kept resolved as the first case (e.g. @page/function always resolved into ./page-a/function subsequently)?

Sorry, I don't undestand, can you clarify?

**Comment by Abdillah** [comment]
> Is it possible to rewrite imports?

Well, no AFAIK. I mean somehow I can workaround the webpack and resolver, or best my project :smile:.

> Sorry, I don't undestand, can you clarify?

I found that Webpack can resolve a path (e.g. `@page/function`) into its alias (e.g. `./page-a/function`). Once a path is resolved, the Webpack always use this resolved value. The second time Webpack find a similar path, it uses again and again. 

I wonder, where do in the source code Webpack store this resolved value?

I thought initially `ResolverCachePlugin` is the one that store this resolved value. But after further inspection, I'm now in doubt.

**Comment by alexander-akait** [comment]
This is our resovler - https://github.com/webpack/enhanced-resolve, pseudo logic is
```js
let result = cache.get(request);

if (!cachedResolved) {
  result = require("enhanced-resolve").resolve(request);
  
  cache.set(request, result);
}


return result
```

That is why I said it is expected, Adding additional cache layers will have a negative perfomance effects and it is unlikely

**Comment by Abdillah** [comment]
I see.. So the `!cachedResolved` kind of condition is inside the `enhanced-resolve`, thanks!
May I ask where is the condition specifically inside the `enhanced-resolve` repository?

**Comment by alexander-akait** [comment]
@Abdillah No, it is inside webpack, it is inside `ResolverCachePlugin` using hooks, why you need this?

**Comment by Abdillah** [comment]
I see. Something's off with my test previously.

I'm curious if I can bypass it, like mentioned above. I'm not going to add additional cache layer, it is rather to bypass it.

**Comment by alexander-akait** [comment]
> I'm curious if I can bypass it, like I mentioned above. I'm not going to add additional cache layer, it is rather to bypass it.

I am afraid no, only if you disable cache
